#159204725 user follow and unfollow author

Oluwafayokemi opened the pull request:
#### What does this PR do?
It ensures that a user can follow and unfollow another user
#### Description of Task to be completed?
- Write unit tests
- An authenticated user should be able to follow an author
- An authenticated user should be able to retrieve the list of all the author he follows
- An authenticated user should be able to retrieve the list of all those who follow him
- An authenticated user should be able to unfollow any of the authors he follows
#### How should this be manually tested?
On postman 
- Make a POST request to '/api/user/follow/:id' - to follow an author
- Make a GET request to '/api/user/follower' - to get the list of all your followers
- Make a GET request to '/api/user/following' - to get the list of the authors you currently follow
#### What are the relevant pivotal tracker stories?
[#159204725](https://www.pivotaltracker.com/story/show/159204725)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/33626790/44409816-00461500-a55b-11e8-8dab-bcc7889044fa.png)
![image](https://user-images.githubusercontent.com/33626790/44409717-d55bc100-a55a-11e8-969a-54c1fd7c7c23.png)
![image](https://user-images.githubusercontent.com/33626790/44409732-dc82cf00-a55a-11e8-991b-0a7d34c3840a.png)

Commit:
create user follow model and relationship

Commit:
feat(follow): implement Users should be able to follow each other
- ensure authenticated user can follow another user
- ensure authenticated user can get the list of those he follows
- ensure authenticated user can get the list of his followers
- ensure authenticated user can unfollow any author
[Delivers #159204725]

Commit:
feat(follow): implement Users should be able to follow each other
- write unit tests
- ensure authenticated user can follow another author
- ensure authenticated user can get the list of those he follows
- ensure authenticated user can get the list of his follower:
- ensure authenticated user can unfollow the author he follows
[Delivers #159204725]

Veraclins reviewed (changes requested):
@Oluwafayokemi, kudos on a very neat job. This is quite easy to follow and apt. There are a few places I believe you may want to look into.
1. The FollowController was exported as ArticleController, I am sure that is a typo. It works because it is a default export that can be referenced by any name but I am sure you will want to export it with the right name.
2. In line 21 - 26 you are checking to make sure that a user does not try to follow him/herself, but putting it below line 19 means there is a db call even if the user is the one attempting the operation. Taking these lines above the db call will save a wasted db call and improve performance.
3. The format of your responses does not follow the Jsend format we agreed to use at the beginning of the project nor any known standard. I think it will help with consistency if you use the agreed format. Responding with "success": false also doesn't make much sense to me.
4. And I noticed a few typos
- line 24: cannot is a single word
- line 57: hold on are two words
- line 90: the key follower should be plural
- line 89: these should be followed by plural users
- line 127: same as line 89 above
- line 128: since you used followee in the method name, the key should be followees or vise versa.
5. Why are the followerId and followingId in the Follow model nullable? Can there be a following without those fields? I feel you should make them not nullable to prevent invalid records in the db.
6. Finally, I think you should make up your mind on whether to use following or followee and stick to it, using the two of them interchangeably is confusing and could lead to errors. 
Kudos!

Commit:
Ti

Commit:
fix follow model to not null

Oluwafayokemi commented:
Hi @veracity, thank you for the feedback. On Line 21-26 I am not making any
extra call to the db with that function, I just had it inside that function
for neatness. Bt it is not redundant. I have implemented the Jsend format
and fixed the typo errors. However, the reason why I am using
getAuthorsFollowee
as the function name is that getAuthorsFollwowing does not sound meaningful
to me. Overall, thank you for the feedback. You are deeply appreciated.

On Thu, Aug 23, 2018 at 12:40 PM Veraclins <notifications@github.com> wrote:

> *@Veraclins* requested changes on this pull request.
>
> @Oluwafayokemi <https://github.com/Oluwafayokemi>, kudos on a very neat
> job. This is quite easy to follow and apt. There are a few places I believe
> you may want to look into.
>
>    1. You FollowController was exported as ArticleController, I am sure
>    that is a typo. It works because it is a default export that can be
>    referenced by any name but I am sure you will want to export it with the
>    right name.
>    2. In line 21 - 26 you are checking to make sure that a user does not
>    try to follow him/herself, but putting it below line 19 means there is a db
>    call even if the user is the one attempting the operation. Taking these
>    lines above the db call will save a wasted db call and improve performance.
>    3. The format of your responses does not follow the Jsend format we
>    agreed to use at the beginning of the project nor any known standard. I
>    think it will help with consistency if you use the agreed format.
>    Responding with "success": false also doesn't make much sense to me.
>    4. And I noticed a few typos
>
>
>    - line 24: cannot is a single word
>    - line 57: hold on are two words
>    - line 90: the key follower should be plural
>    - line 89: these should be followed by plural users
>    - line 127: same as line 89 above
>    - line 128: since you used followee in the method name, the key should
>    be followees or vise versa.
>
>
>    1. Why are the followerId and followingId in the Follow model
>    nullable? Can there be a following without those fields?
>    2. Finally, I think you should make up your mind on whether to use
>    following or followee and stick to it, using the two of them
>    interchangeably is confusing and can lead to errors.
>    Kudos!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/andela/elven-ah/pull/24#pullrequestreview-148870016>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgEappZP3ZuVGXiy02Se6PF0me8qKuRPks5uTpTBgaJpZM4WF9f7>
> .
>

Commit:
fix: prevent extra db call while verifying authorsId

Oluwafayokemi commented:
Thank you @veracity, I understand you better now. Thank you for clarifying
yourself on line 21-26. I have implemented your feedback. Thanks


On Thu, Aug 23, 2018 at 2:49 PM Fayokemi Adeyina <fayoaright@gmail.com>
wrote:

> Hi @veracity, thank you for the feedback. On Line 21-26 I am not making
> any extra call to the db with that function, I just had it inside that
> function for neatness. Bt it is not redundant. I have implemented the Jsend
> format and fixed the typo errors. However, the reason why I am using getAuthorsFollowee
> as the function name is that getAuthorsFollwowing does not sound
> meaningful to me. Overall, thank you for the feedback. You are deeply
> appreciated.
>
> On Thu, Aug 23, 2018 at 12:40 PM Veraclins <notifications@github.com>
> wrote:
>
>> *@Veraclins* requested changes on this pull request.
>>
>> @Oluwafayokemi <https://github.com/Oluwafayokemi>, kudos on a very neat
>> job. This is quite easy to follow and apt. There are a few places I believe
>> you may want to look into.
>>
>>    1. You FollowController was exported as ArticleController, I am sure
>>    that is a typo. It works because it is a default export that can be
>>    referenced by any name but I am sure you will want to export it with the
>>    right name.
>>    2. In line 21 - 26 you are checking to make sure that a user does not
>>    try to follow him/herself, but putting it below line 19 means there is a db
>>    call even if the user is the one attempting the operation. Taking these
>>    lines above the db call will save a wasted db call and improve performance.
>>    3. The format of your responses does not follow the Jsend format we
>>    agreed to use at the beginning of the project nor any known standard. I
>>    think it will help with consistency if you use the agreed format.
>>    Responding with "success": false also doesn't make much sense to me.
>>    4. And I noticed a few typos
>>
>>
>>    - line 24: cannot is a single word
>>    - line 57: hold on are two words
>>    - line 90: the key follower should be plural
>>    - line 89: these should be followed by plural users
>>    - line 127: same as line 89 above
>>    - line 128: since you used followee in the method name, the key
>>    should be followees or vise versa.
>>
>>
>>    1. Why are the followerId and followingId in the Follow model
>>    nullable? Can there be a following without those fields?
>>    2. Finally, I think you should make up your mind on whether to use
>>    following or followee and stick to it, using the two of them
>>    interchangeably is confusing and can lead to errors.
>>    Kudos!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/andela/elven-ah/pull/24#pullrequestreview-148870016>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AgEappZP3ZuVGXiy02Se6PF0me8qKuRPks5uTpTBgaJpZM4WF9f7>
>> .
>>
>

yomigeek commented on server/controllers/FollowController.js:
```diff
@@ -0,0 +1,171 @@
+   * @param {object} res the response object
+   * @returns {object} list of all user author the author follows.
+   */
+  static getAuthorsFollowee(req, res) {
```
Good work @Oluwafayokemi. I feel you could use getUserFollowee instead of getAuthorsFollowee, considering the fact that authors and readers are all users.

madeofhuman reviewed (approved):
@Veraclins has said almost everything there is to be said (you were really quite thorough, man). Good job, @Oluwafayokemi

yomigeek reviewed (approved):
Good work @Oluwafayokemi .. I must say I commend your effort and swiftness in completing the task.
